Add a build flag to use left folds for elimination ops in stream module

harendra-kumar opened the pull request.

adithyaov commented on core/src/Streamly/Internal/Data/Fold.hs:
```diff
@@ -1166,17 +1165,25 @@ head = one
-        if predicate a
-        then Done (Just a)
-        else Partial ()
+        (\r ->
+            if r
+            then Done (Just a)
+            else Partial ()) <$> predicate a
```
Any reason for this style of code?
Maybe move the function to a where or a let statement?
This is rather hard to read.

harendra-kumar replied:
reason: least amount of change from previous code. Yes, can define the func separately.

harendra-kumar commented:
Benchmark difference without `use-folds` flag and with it. No difference in allocations.

```
Prelude.Serial(cpuTime)
Benchmark                                                                                                     default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------------------------------------- -------------- --------------------------
All.Prelude.Serial/o-1-space.concat.concatMapRepl (sqrt n of sqrt n)                                                  740.61                     +18.40
All.Prelude.Serial/o-1-space.filtering.dropWhile-true                                                                  47.37                     +18.15
All.Prelude.Serial/o-1-space.filteringX4.findIndices                                                                   65.76                     +13.29
All.Prelude.Serial/o-1-space.exceptions.S.after . S.before (1/10)                                                   10002.00                     +13.27
All.Prelude.Serial/o-1-space.split.S.splitOnSeq "abcdefghi" FL.drain                                               362021.00                     +11.21
All.Prelude.Serial/o-1-space.exceptions.S.after (1/10)                                                              10005.70                     +10.76
All.Prelude.Serial/o-1-space.elimination.findM                                                                         59.61                      +9.57
All.Prelude.Serial/o-1-space.split.S.splitOnSeq "catcatcatcatcat" FL.drain                                         358347.00                      +9.20
All.Prelude.Serial/o-1-space.concat.concatMap (1 of n)                                                               1491.83                      +9.20
All.Prelude.Serial/o-n-heap.toList.toListRev                                                                         6206.52                      +7.73
All.Prelude.Serial/o-1-space.split.S.splitOnSeq "abcdefghijklmnopqrstuvwxyz" FL.drain                              382034.00                      +7.59
All.Prelude.Serial/o-n-stack.iterated.dropOne (n/10 x 10)                                                             243.94                      +7.03
All.Prelude.Serial/o-1-space.mapping.timestamped                                                                       82.27                      +6.90
All.Prelude.Serial/o-n-heap.buffered.reverse                                                                         6220.18                      +6.78
All.Prelude.Serial/o-1-space.elimination.drainN                                                                        55.98                      +6.65
All.Prelude.Serial/o-n-heap.foldl.foldlM'/build/Identity                                                             7606.81                      +6.57
All.Prelude.Serial/o-n-space.toList.toStream                                                                        10404.50                      +6.11
All.Prelude.Serial/o-1-space.split.S.foldMany (FL.takeEndBy_ (== lf) FL.drain)                                      78066.70                      +5.58
All.Prelude.Serial/o-n-heap.foldl.foldl'/build/IO                                                                    6302.21                      +5.52
All.Prelude.Serial/o-1-space.elimination.length . generally                                                            47.20                      +5.50
All.Prelude.Serial/o-1-space.elimination.build.IO.foldrMElem                                                           47.11                      +5.49
All.Prelude.Serial/o-1-space.mapping.postscanlM'                                                                       47.14                      +5.46
All.Prelude.Serial/o-1-space.grouping.foldIterateM                                                                    453.48                      +5.46
All.Prelude.Serial/o-1-space.mapping.scanl1'                                                                           47.17                      +5.39
All.Prelude.Serial/o-1-space.elimination.elemIndex                                                                     56.67                      +5.28

...

All.Prelude.Serial/o-1-space.elimination.length                                                                        49.63                      -3.15
All.Prelude.Serial/o-1-space.foldable.min (ord)                                                                      1066.60                      -3.28
All.Prelude.Serial/o-1-space.Monad.(>>=) (sqrt n x sqrt n)                                                           6251.81                      -3.52
All.Prelude.Serial/o-1-space.split.S.splitOnSuffixSeq "\n" FL.drain                                                 81762.90                      -3.55
All.Prelude.Serial/o-1-space.elimination.minimumBy                                                                     49.75                      -3.74
All.Prelude.Serial/o-1-space.elimination.uncons                                                                      1062.42                      -3.75
All.Prelude.Serial/o-1-space.mapping.foldrSMap                                                                       3244.15                      -3.76
All.Prelude.Serial/o-1-space.generation.fromListM                                                                    1061.62                      -3.92
All.Prelude.Serial/o-1-space.generation.IsList.fromList                                                               788.72                      -4.17
All.Prelude.Serial/o-1-space.generation.absTimes                                                                       92.09                      -4.36
All.Prelude.Serial/o-1-space.elimination.and                                                                           49.74                      -4.58
All.Prelude.Serial/o-1-space.mapping.scanl'                                                                            49.66                      -4.91
All.Prelude.Serial/o-1-space.mapping.postscanl'                                                                        49.73                      -5.11
All.Prelude.Serial/o-1-space.filteringX4.takeWhile-true                                                                49.72                      -5.11
All.Prelude.Serial/o-1-space.mixed.scanl-map                                                                           49.73                      -5.14
All.Prelude.Serial/o-1-space.elimination.reduce.IO.foldl'                                                              49.70                      -5.22
All.Prelude.Serial/o-1-space.Monad.(>>=) (sqrt n x sqrt n) (filterAllOut)                                            5783.75                      -5.24
All.Prelude.Serial/o-n-heap.buffered.joinInner (sqrtVal)                                                             4324.27                      -5.26
All.Prelude.Serial/o-1-space.exceptions/serial.retryNone                                                             1662.65                      -5.44
All.Prelude.Serial/o-1-space.mapping.scanl1M'                                                                          49.76                      -5.77
All.Prelude.Serial/o-1-space.filteringX4.takeWhileM-true                                                               49.74                      -5.77
All.Prelude.Serial/o-n-heap.grouping.classifySessionsOfHash (10000 buckets)                                        188512.00                      -5.81
All.Prelude.Serial/o-1-space.filteringX4.foldFilter-even                                                             6334.19                      -5.86
All.Prelude.Serial/o-1-space.generation.iterate                                                                       485.19                      -6.56
All.Prelude.Serial/o-1-space.exceptions/serial.retryNoneSimple                                                       1751.92                      -6.64
All.Prelude.Serial/o-1-space.elimination.build.Identity.foldrToStreamLength                                          1107.78                      -6.73
All.Prelude.Serial/o-1-space.grouping.groupsByRollingEq                                                                59.73                      -6.77
All.Prelude.Serial/o-1-space.filtering.foldFilter-even                                                                129.31                      -6.77
All.Prelude.Serial/o-1-space.elimination.init                                                                        3234.00                      -7.26
All.Prelude.Serial/o-1-space.elimination.reduce.Identity.foldl1'                                                     1108.52                      -7.58
All.Prelude.Serial/o-1-space.elimination.reduce.IO.foldl1'                                                            999.65                      -7.83
All.Prelude.Serial/o-1-space.mixedX4.take-scan                                                                        223.51                      -8.91
All.Prelude.Serial/o-1-space.generation.IsString.fromString                                                           682.31                     -10.95
All.Prelude.Serial/o-1-space.elimination.all                                                                           55.88                     -12.28
All.Prelude.Serial/o-1-space.split.S.splitOnSeq "abcdefgh" FL.drain                                                173255.00                     -13.50
All.Prelude.Serial/o-n-space.foldl.minimumBy                                                                        19430.90                     -14.72
All.Prelude.Serial/o-1-space.split.S.parseMany (FL.takeEndBy_ (== lf) FL.drain)                                    115485.00                     -15.04
All.Prelude.Serial/o-n-heap.transformer.StateT Int IO (n times) (baseline)                                          10998.10                     -15.60
All.Prelude.Serial/o-1-space.elimination.maximum                                                                       65.25                     -16.40
All.Prelude.Serial/o-1-space.elimination.reduce.Identity.foldl'                                                        86.95                     -16.66
All.Prelude.Serial/o-1-space.elimination.elem                                                                          59.57                     -19.72
All.Prelude.Serial/o-1-space.elimination.(!!)                                                                          74.58                     -31.71
```

Seems no significant difference. We can make this default.
